x-pack/plugins/cloud/server/plugin.ts

```diff
@@ -202,7 +202,7 @@ export class CloudPlugin implements Plugin<CloudSetup, CloudStart> {
       organizationId,
       instanceSizeMb: readInstanceSizeMb(),
       deploymentId,
-      elasticsearchUrl: decodedId?.elasticsearchUrl,
+      elasticsearchUrl: core.elasticsearch.publicBaseUrl || decodedId?.elasticsearchUrl,
```

Comment:
for `publicBaseUrl` to be useful for users outside of cloud (e.g. when running kibana/es in docker containers with #start-local) we need to move all consumers of `CloudSetup.elasticsearchUrl` to `core.elasticsearch.publicBaseUrl`.

But we can only do that once we know that the `elasticsearch.publicBaseUrl` config is set correctly by control plane. So I think we'd want to do this in phases:

1. Merge this PR to add the kibana.yml `elasticsearch.publicBaseUrl` option.
2. Update elastic/kibana-controller to populate this value on serverless (wait for serverless release of Kibana before merging!)
3. Update elastic/cloud to populate this value for ECH when Kibana >= 8.16 (or whichever release we land in)
4. Remove `elasticsearchUrl` from Cloud plugin and change all consumers to usecore.elasticsearch.publicBaseUrl` instead

Reply:
Yup, that sequence makes sense to me. I'll merge this PR and then we can get started on the next steps.

Reply:
JIRA issue to cover step 2: https://elasticco.atlassian.net/browse/CP-6363
JIRA issue to cover step 3: https://elasticco.atlassian.net/browse/CP-8356